# Single-nucleus resolution of sex-biased expression and dosage compensation in *Drosophila melanogaster*

### Here is a repository that includes all R scripts used throughout our analysis of the Fly Cell Atlas dataset (Li et al, 2022).

| **Script**                                             | **Required**       | **Order** | **Description**                                                                                                                                                                                                                                                                                                                                                                                                                                                   |
|--------------------------------------------------------|--------------------|-----------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **remove_sex_specific_clusters_from_nonrep_tissues.R** | Analysis           | 1         | Convert loom format to h5Seurat and remove sex-specific clusters from somatic tissues                                                                                                                                                                                                                                                                                                                                                                             |
| **run_sctransform_on_r_fca_raw_filtered.R**            | Analysis           | 2         | Run stringent filtering on cells and genes, remove ambient RNA with DecontX and perform normalisation step with sctransform for each tissue                                                                                                                                                                                                                                                                                                                       |
| **de_analysis_with_deseq2_on_head_and_body.R**         | Analysis           | 3         | Run DESeq2 on aggregated counts (pseudo-bulk) in head and carcass at the tissue- and cluster-level                                                                                                                                                                                                                                                                                                                                                                |
| **de_analysis_deseq2_plotting_results.R**              | Plots              | 4         | Plot DESeq2 results at the tissue- and cluster-level                                                                                                                                                                                                                                                                                                                                                                                                              |
| **de_analysis_calculate_sex_bias2.R**                  | Analysis           | 5         | Calculate sex bias with and without accounting for cell type composition differences between the sexes                                                                                                                                                                                                                                                                                                                                                            |
| **plotting_sex_bias_results.R**                        | Plots              | 6         | Plot sex bias in cell type proportion and gene expression for each tissue                                                                                                                                                                                                                                                                                                                                                                                         |
| **x_vs_a_ratio_analysis.R**                            | Analysis           | 7         | Calculate mean X:A expression and obtain expression of DCC-related genes for each cell                                                                                                                                                                                                                                                                                                                                                                            |
| **plotting_x-to-a_expression_results.R**               | Plots              | 8         | Plot X:A expression ratios across tissues and cell types, and in the male gonad                                                                                                                                                                                                                                                                                                                                                                                   |
| **investigating_male_reproductive_glands_2.R**         | Plots              | 9         | Plot X:A expression ratios across male accessory gland cell type clusters; plot  X:A for each cell type cluster (ridges and UMAP); plot roX genes and Sxl expression in the accessory gland; dotplot of DCC-related in male accessory gland cell type clusters; plot percentage of autosomal and X-linked genes expressed in the accessory gland; dotplot of germline markers in sperm cells and in the male accessory glands; plot dsx expression across tissues |
| **majane_et_el_data_analysis.R**                       | Analysis and plots | 10        | Convert Majane et al (2022) data into h5Seurat format, stringent filtering, ambient RNA removal with DecontX followed by normalisation with sctransform, and calculate X:A expression for each cell within the male accessory gland; plot X:A expression for each cell type cluster (ridges and UMAP), and generate dotplots of DCC-related genes for each cluster in the male accessory gland                                                                    |
| **majane_et_el_differential_expression.R**             | Analysis           | 11        | Perform a DE analysis between main cell clusters in the Majane et al (2022) dataset using MAST                                                                                                                                                                                                                                                                                                                                                                    |
| **plotting_bulk_vs_pseudo_bulk_rnaseq.R**              | Analysis and plots | 12        | Calculate and plot average expression in bulk RNAseq (from Leader et al, 2017) versus pseudo-bulk (Fly Cell Atlas data) for each gene found in both datasets                                                                                                                                                                                                                                                                                                      |
