# Intermediate files in our study

Description of all the intermediate files generated and used throughout our study.

| **File number** | **Filename**                                                                                               | **Description**                                                                                                                                                |
|-----------------|------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 1               | cell_types_and_broad_annotation2.csv                                                                       | Metadata describing Fly Cell Atlas' annotated cell-type clusters                                                                                               |
| 2               | deseq_results_from_raw_relaxed_body_per_cluster.csv                                                        | DESeq2 results across cell-type clusters in the soma                                                                                                           |
| 3               | deseq_results_from_raw_relaxed_body_per_tissue.csv                                                         | DESEq2 results in the soma                                                                                                                                     |
| 4               | deseq_results_from_raw_relaxed_head_per_cluster.csv                                                        | DESeq2 results across cell-type clusters in the head                                                                                                           |
| 5               | deseq_results_from_raw_relaxed_head_per_tissue.csv                                                         | DESeq2 results in the head                                                                                                                                     |
| 6               | observed_over_expected_sex_biased_genes_autosomes_and_xchr_per_cluster.csv                                 | Number of observed over expected sex-biased genes based on number of autosomal and X-linked genes for each cell-type cluster                                   |
| 7               | fet_sex_biased_genes_autosomes_and_xchr_per_cluster.csv                                                    | Results of Fisher's exact test that tests whether the proportion of sex-biased genes differs between the X chromosome and autosomes for each cell-type cluster |
| 8               | sex_bias_cell_vs_no_cell_per_tissue_plus.zip                                                               | Computed values for sex-bias (not) accounting for sex differences in cell-type composition across cell-type clusters                                           |
| 9               | number_of_cells_per_cluster_all_tissues.csv                                                                | Number of cells per cluster across tissues                                                                                                                     |
| 10              | x_a_normalised_expression_ratio_n_max_expression_data_relaxed_after_filtering_all_tissues.zip              | Average X:A expression values and MSL-related genes counts across tissues                                                                                      |
| 11              | s_testis_umap_reduction_coordinates.csv                                                                    | UMAP coordinates from the FCA for cells in the testis                                                                                                          |
| 12              | x_a_normalised_expression_ratio_n_max_expression_data_male_reproductive_glands_relaxed_after_filtering.csv | Average X:A expression values and MSL-related genes counts in the male accessory glands                                                                        |
| 13              | mast_result_mag_main_cell_clusters.txt                                                                     | Differential expression results using MAST to compare main cell clusters in the MAG                                                                            |
| 14              | fca_male_germline_cell_markers.csv                                                                         | Marker genes for male germline cells                                                                                                                           |
| 15              | x_a_normalised_expression_ratio_n_max_og_clean_expression_data_majane_etal.csv                             | Average X:A expression values and MSL-related genes counts in the Dmel MAG from Majane et al (2022)                                                            |
| 16              | x_a_normalised_expression_ratio_n_max_og_clean_sim_expression_data_majane_etal.csv                         | Average X:A expression values and MSL-related genes counts in the Dsim MAG from Majane et al (2022)                                                            |
| 17              | x_a_normalised_expression_ratio_n_max_og_clean_yak_expression_data_majane_etal.csv                         | Average X:A expression values and MSL-related genes counts in the Dyak MAG from Majane et al (2022)                                                            |
| 18              | majane_etal_main_cell_markers.csv                                                                          | Marker genes for main cell 1 and 2 clusters in Majane et al (2022)                                                                                             |
| 19              | majane_dmel_mast_result_mag_main_cell_clusters.txt                                                         | Differential expression results using MAST to compare main cell clusters in the Dmel MAG of Majane et al (2022)                                                |
